docs/configuration.rst
.. _configuration:

************************************************************************
Configuration
************************************************************************

.. contents:: :local:

General
------------------------------------------------------------------------

``err-stackstorm`` configuration is found in Errbot's ``config.py`` file.

.. note:: If you followed the Errbot setup documentation this file will have been created by downloading a template.  If this file has not already been created, please create it following the `Errbot's instructions <http://errbot.io/en/latest/user_guide/setup.html#id1>`_.

Here's a sample err-stackstorm configuration:

.. code-block:: python

    STACKSTORM = {
        "auth_url": "https://your.stackstorm.com/auth/v1",
        "api_url": "https://your.stackstorm.com/api/v1",
        "stream_url": "https://your.stackstorm.com/stream/v1",
        "route_key": "errbot",
        "plugin_prefix": "st2",
        "verify_cert": True,
        "secrets_store": "cleartext",
        "session_ttl": 3600,
        "user_token_ttl": 86400,
        "api_auth": {
            "user": {
                "name": "my_username",
                "password": "my_password",
            },
            "token": "<User token>",
            "apikey": "<API Key>"
        },
        "rbac_auth": {
            "standalone": {},
        },
        "timer_update": 900, #  Unit: second.  Interval to check the user token is still valid.
    }

ST2 ChatOps Configuration
------------------------------------------------------------------------

StackStorm's `ChatOps pack <https://github.com/StackStorm/st2/tree/master/contrib/chatops>`_ has to be installed and a notify rule file added to the pack.

The notify rule must be placed in ``/<st2>/packs/chatops/rules``.  You can find the rule file necessary for `err-stackstorm` `here <https://raw.githubusercontent.com/nzlosh/err-stackstorm/master/contrib/stackstorm-chatops/rules/notify_errbot.yaml>`_.  Just copy-and-paste this file and save it as ``notify_errbot.yaml``.

Edit the ``/<st2>/packs/chatops/actions/post_message.yaml`` file and replace `chatops` with ``errbot``:

.. code-block:: yaml

  route:
    default: "errbot"

.. note:: See Route Key below for information on customising the route.

Authentication
------------------------------------------------------------------------

`err-stackstorm` must have valid credentials to use StackStorm's API.
StackStorm's authentication is possible with:

* Username and password
* User token
* API key

See `StackStorm's Authentication docs <https://docs.stackstorm.com/authentication.html>`_ for more details on how to generate credentials.

Username/Password
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Using a username and password will allow `err-stackstorm` to renew the user token once it expires.

.. note:: If you specify both username/password and a User Token, the User Token will be used and the *username/password will be ignored*.

User Token
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To avoid using the username/password pair in a configuration file, it's possible to supply a User Token as generated by StackStorm when a username/password is authenticated successfully.

If using an User Token `err-stackstorm` will no longer have access to the StackStorm's API once it expires.

.. note:: When the token expires, a new one must be generated and updated in config.py which in turn requires Errbot to be restarted.  This form of authentication is the least practical for production environments.

API Key
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*API Key support has been included since StackStorm v2.0.*

When an API Key is provided, all other authentication settings are ignored and the API Key is used.

If using an API Key, `err-stackstorm` will no longer have access to the StackStorm's API once it expires.

.. note:: It is considered a mistake to supply a token or username/password pair when using the API Key.

Secrets Store
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The secrets store is used by `err-stackstorm` to cache StackStorm API credentials.  The available backends are:

* cleartext


Cleartext
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

The cleartext store maintains the cache in memory and does not encrypt the contents to disk.  It **does not** protect the stored secrets in memory.

Advanced Options
------------------------------------------------------------------------

Route Key
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The route key is used by err-stackstorm to inform StackStorm where to send result notifications for action-aliases.  StackStorm sends notification events via the stream interface that are marked with the route key.  Err-stackstorm filters these events using the route key and will handle any events that match its configured route key.

By altering the route key, it is possible to have multiple instances of err-stackstorm that are connected to the same StackStorm instance.  This would allow for configurations where StackStorm is available on multiple chat backends.

Example
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
A StackStorm instance has 2 err-stackstorm instances.  The first err-stackstorm instance uses the `errbot-slack` route key while the second instance uses the `errbot-discord` route key.  Both instances have the same plugin prefix and expose the same action-aliases.  This would mean the command `!st2 pack list` could be run on discord and the result notification would be routed to the `errbot-discord` err-stackstorm instance.


Plugin Prefix
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default the plugin prefix is set to `st2`.  The plugin prefix serves to prevent action-aliases collisions with errbot's native plugins commands.  It is possible to customise the plugin prefix to use a more appropriate naming scheme for the environment err-stackstorm is running.

Aside from cosmetic value, customising the plugin prefix can allow for multiple err-stackstorm instances to occupy the same chat channel.  This would be achieved by setting a unique plugin-prefix per instance.

.. note:: Always use strings that do not conflict with existing errbot commands.

Example
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
In the case of multiple instances of a StackStorm and err-stackstorm pair, say 1 per data centre or 1 per region, it would be possible to assign a unique plugin prefix per instance.  Let's say there were 2 pairs with the plugin-prefix of `dc1` and `dc2`.  Both bots could occupy `#automation` channel and users could trigger the pack list action-alias in data centre #1 by calling `!dc1 pack list` or get available action-aliases from data centre #2 by calling `!dc2help`.

Session TTL
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The session time to live is an internal timeout for err-stackstorm.  It is used to set the maximum lifetime an err-stackstorm session is permitted to exist.  Once the session expires the user will need to re-authenticate before being able to execute action-alias commands.  An err-stackstorm session ttl should be equal to or less than the user token ttl.

When session time to live expires, err-stackstorm will report to the user that the session is no longer valid and they should re-authenticate.

User Token TTL
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The user token ttl is used to set the maximum life time a StackStorm User Token will be permitted to exist.  Once the user token has expired the user will need to re-authenticate before being able to execute action-alias commands.

When a user token time to live expires, err-stackstorm will report it as an error communicating with the StackStorm API.  It would be more user-friendly to ensure the session ttl expires before the user token ttl.

.. note:: The user token ``ttl`` *must be* equal to or lower than the StackStorm API https://docs.stackstorm.com/authentication.html?highlight=ttl#usage.  By default StackStorm's token ttl is set to 24 hours, but the value can be increased through ``st2.conf``.  If user_token_ttl is greater than the StackStorm API token ttl value, err-stackstorm will fail to fetch a valid API token and not function correctly.

Locale
------------------------------------------------------------------------

Errbot uses the system's locale for handling text.  If you're getting unicode errors like this::

  UnicodeEncodeError: 'ascii' codec can't encode character '\xe9' in position 83: ordinal not in range(128)

Make sure the systems locale is configured for unicode encoding.  In the example below, the machine has been set to English (en) New Zealand (NZ) with utf-8 encoding (.UTF8).

.. code-block:: bash

  # locale
  LANG=en_NZ.UTF8
  LANGUAGE=
  LC_CTYPE="en_NZ.UTF8"
  LC_NUMERIC="en_NZ.UTF8"
  LC_TIME="en_NZ.UTF8"
  LC_COLLATE="en_NZ.UTF8"
  LC_MONETARY="en_NZ.UTF8"
  LC_MESSAGES="en_NZ.UTF8"
  LC_PAPER="en_NZ.UTF8"
  LC_NAME="en_NZ.UTF8"
  LC_ADDRESS="en_NZ.UTF8"
  LC_TELEPHONE="en_NZ.UTF8"
  LC_MEASUREMENT="en_NZ.UTF8"
  LC_IDENTIFICATION="en_NZ.UTF8"
  LC_ALL=en_NZ.UTF8


Reference
------------------------------------------------------------------------


  .. csv-table::
    :header: "Option", "Description"
    :widths: 25, 40

    "auth_url", "StackStorm's authentication url end point.  Used to authenticate credentials against StackStorm."
    "api_url", "StackStorm's API url end point.  Used to execute action aliases received from the chat back-end."
    "stream_url", "StackStorm's Stream url end point.  Used to received ChatOps notifications."
    "verify_cert", "Default is *True*.  Verify the SSL certificate is valid when using https end points.  Applies to all end points."
    "route_key", "Default is *errbot*.  The name of the route to bot will listen for and submit action-alias executions with."
    "plugin_prefix", "Default is *st2*.  Text used to prefix action-alias commands with to avoid name collisions between StackStorm Action-Aliases and Errbot plugin commands."
    "api_auth.user.name", "Errbot's username to authenticate with StackStorm."
    "api_auth.user.password", "Errbot's password to authenticate with StackStorm."
    "api_auth.token", "Errbot's user token to authenticate with StackStorm.  Used instead of a username/password pair."
    "api_auth.apikey", "Errbot API key to authenticate with StackStorm.  Used instead of a username/password pair or user token."
    "timer_update", "Unit: seconds.  Default: 60.  Interval for err-stackstorm to the user token is valid."
    "rbac_auth.standalone", "Standalone authentication."
    "rbac_auth.clientside", "Clientside authentication, a chat user will supply StackStorm credentials to err-stackstorm via an authentication page."
    "rbac_auth.clientside.url", "Url to the authentication web page."
    "session_ttl", "Unit: seconds.  Default: 3600.  The time to live for a authentication session."
    "user_token_ttl", "Unit: seconds.  Default: 86400.  The time to live for a StackStorm user token."
    "secrets_store.cleartext", "Use the in-memory store."



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

##  [Unreleased]
### Added
### Changed
### Removed

##  [2.2.0] 2021-11-27
### Added
  - support for slackv3 backend and passing action-alias extra parameters as Slack block or attachment.
  - documentation to use Slack blocks or attachments.
  - tests for Python from 3.6 to 3.10

### Changed
  -  hard coded notification_route to use user configured route when calling match_and_execute.

### Removed
  - logging sensitive api tokens in debug logs.
  - "<action-alias>" for action execution help text.

##  [2.1.4] 2020-08-14
### Added
  - Session are deleted automatically when a chat user fails to authenticate against St2 API.
  - Include Slack users "display name" in session list for easier identification.

### Changed
  - Update documentation with corrections and improved examples.
  - Corrected error of dynamic commands not registering correctly when the plugin was deactivated/activated.
  - Improved robustness of version check code to gracefully handle github.com being unavailable.

##  [2.1.3] 2020-01-10
### Added
  - Added err-stackstorm version in the st2help display.

### Changed
  - Fixed st2help function to have to correct number of positional arguments.

##  [2.1.2] 2019-11-24
### Added
  - Added CircleCI build badge to README.
  - Added route key to be defined by user to allow mutliple bots attached to a single StackStorm instance.
  - Added version check on start-up to log if newer version of the plugin are available.

### Changed
  - Updated curl example in troubleshooting section.
  - Changed all bot commands to be registered dynamically to allow user defined plugin prefix.

##  [2.1.1] 2019-07-19
### Added
  - Added configuration for mattermost, rocketchat, discord, gitter and slack to docker build.

### Changed
  - Improved Discord Chat Adapter formatting.
  - Corrected configuration name handling which caused exceptions to be raised.
  - Updated documentation to direct readers to readthedocs.

### Removed
  - Removed old readme file.

## [2.1.0] 2019-07-08
### Added
  - Added `deactivate` method to correctly handle errbot reload and stop process.
  - Added initial files for documentation to be served on readthedocs.
  - Added missing documentation for nginx configuration when using Client-side authentication.
  - Added support for IRC and Discord Chat Adapters.
  - Added source documentation for ReadTheDocs.

### Changed
  - Correctly detect when the `extra` attribute is passed to the Slack Chat Adapter.
  - Send web login form data as JSON to be compatible with errbot >=v6.0.

### Removed

## [2.1.0-rc2] 2019-05-29
### Added
  - Split unit tests in separate test files.
  - Added linting to CircleCI
  - Added try/except blocks around plugin configuration code to improve installation and startup
    experience when there are errors in the configuration.

### Changed
  - Fixed dictionary keys to reference 'apikey' consistently.
  - Switch CircleCI from calling pytest directly to using Makefile.
  - Dropped using sseclient-py in favour of btubbs sseclient.
  - Corrected references to apikey configuration.

### Removed

## [2.1.0-rc1] 2019-05-15
### Added
  - unit tests for more low level objects used by err-stackstorm.

### Changed
  - updated makefile to be better adapted to CircleCI's build process.  (Still a work in progress)
  - numerous documentation updates to improve installation process for new users.
  - documented errbot's ACL features.

### Removed
  - removed unused code for keyring and vault.

## [2.0.0-rc2] 2019-02-08
### Added
  - Correctly support action-alias acknowledge enabled flag in StackStorm API.
  - Support appending the execution `web_url` to action-alias acknowledgement message.

### Changed
  - Fixed exception raise in log message when err-stackstorm isn't able to fetch a valid StackStorm user token.

### Removed
  - removed dependency for st2client to be able to use requests>=2.20.  This avoid conflicts with popular chat backends.

## [2.0.0-rc1] 2019-01-28
### Added
 - setup.py to be able to install plugin from the command line.
 - Dockerfile to build an image and run err-stackstorm from within a Docker container. (alpine linux based)
 - Support for Out-of-Bands authentication.
 - Support for formatting with XMMP backend.

### Changed
 - Updated err-stackstorm python module requirements.

### Removed

## [1.4.0] - 2018-04-07
### Added


### Changed
 - Readme spelling corrections.
 - Aligned `api_url`, `auth_url` and `stream_url` with the [StackStorm API](https://api.stackstorm.com/)


### Removed
 - remove `base_url` and `api_version`, they're now calculated from the `api_url`, `auth_url` or `stream_url`.


lib/config.py
# coding:utf-8
import logging

from lib.authentication_handler import AuthHandlerFactory
from lib.credentials_adapters import CredentialsFactory

LOG = logging.getLogger(__name__)


class BorgSingleton:
    """
    Borg Singleton pattern as described in Python Patterns, Idioms and Tests.
    """

    _shared_state = None

    def __init__(self):
        self.__dict__ = BorgSingleton._shared_state


class PluginConfiguration(BorgSingleton):
    """
    err-stackstorm shared configuration.
    """

    def __init__(self):
        super(BorgSingleton, self).__init__()

    def setup(self, bot_conf):
        if not hasattr(bot_conf, "STACKSTORM"):
            LOG.critical(
                "Missing STACKSTORM configuration in config.py.   err-stackstorm must be configured"
                " correctly to function.  See the err-stackstorm documentation for configuration "
                "instructions."
            )
            bot_conf.__setattr__("STACKSTORM", {})
        self._configure_prefixes(bot_conf)
        self._configure_credentials(bot_conf)
        self._configure_rbac_auth(bot_conf)
        self._configure_stackstorm(bot_conf)
        self.oob_auth_url = bot_conf.STACKSTORM.get("oob_auth_url", "https://localhost:8888/")
        self.timer_update = bot_conf.STACKSTORM.get("timer_update", 60)
        self.verify_cert = bot_conf.STACKSTORM.get("verify_cert", True)
        self.secrets_store = bot_conf.STACKSTORM.get("secrets_store", "cleartext")
        self.route_key = bot_conf.STACKSTORM.get("route_key", "errbot")
        self.session_ttl = bot_conf.STACKSTORM.get("session_ttl", 3600)
        self.user_token_ttl = bot_conf.STACKSTORM.get("user_token_ttl", 86400)

        self.client_cert = bot_conf.STACKSTORM.get("client_cert", None)
        self.client_key = bot_conf.STACKSTORM.get("client_key", None)
        self.ca_cert = bot_conf.STACKSTORM.get("ca_cert", None)

    def _configure_rbac_auth(self, bot_conf):
        self.auth_handler = None
        rbac_auth = bot_conf.STACKSTORM.get("rbac_auth", {"standalone": {}})
        for rbac_type in list(rbac_auth.keys()):
            self.auth_handler = AuthHandlerFactory.instantiate(rbac_type)(
                self, rbac_auth[rbac_type]
            )
        if self.auth_handler is None:
            LOG.warning(
                "Failed to configure RBAC authentication handler.  Check the configuration."
            )
        return True

    def _configure_prefixes(self, bot_conf):
        self.bot_prefix = bot_conf.BOT_PREFIX
        self.plugin_prefix = bot_conf.STACKSTORM.get("plugin_prefix", "st2")
        self.full_prefix = "{}{} ".format(bot_conf.BOT_PREFIX, self.plugin_prefix)

    def _configure_stackstorm(self, bot_conf):
        self.api_url = bot_conf.STACKSTORM.get("api_url", "http://localhost:9101/v1")
        self.auth_url = bot_conf.STACKSTORM.get("auth_url", "http://localhost:9100/v1")
        self.stream_url = bot_conf.STACKSTORM.get("stream_url", "http://localhost:9102/v1")

    def _configure_credentials(self, bot_conf):
        self.api_auth = bot_conf.STACKSTORM.get("api_auth", {})
        self.bot_creds = None
        for cred_type in ["apikey", "user", "token"]:
            c = self.api_auth.get(cred_type)
            if c:
                if cred_type == "user":
                    self.bot_creds = CredentialsFactory.instantiate(cred_type)(
                        c.get("name"), c.get("password")
                    )
                    break
                else:
                    self.bot_creds = CredentialsFactory.instantiate(cred_type)(c)
                    break
        if self.bot_creds is None:
            LOG.warning(
                "Failed to find any valid st2 credentials for the bot,  check configuration file."
            )
